server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -267,9 +264,13 @@ public ObjectMapper findNestedObjectMapper(int nestedDocId, SearchContext sc, Le
         return nestedObjectMapper;
     }
 
-    public DocumentMapper merge(Mapping mapping, MergeReason reason) {
+    public DocumentMapper merge(Mapping mapping,
+                                MergeReason reason,
+                                IndexSettings indexSettings,
```

Comment:
Rather than having to pass these through to `merge`, can't we just keep references to them instead of references to the MapperService?

Reply:
I don't see the point in holding on to references of objects that callers also hold on to already? Also, looking at potential further improvements, I think not holding those references simplifies things in DocumentMapper, doesn't it? are you annoyed by the number of arguments required by calling merge?

Reply:
The IndexSettings, IndexAnalyzers and DocumentMapperParser always remain the same, I think? So it's annoying to have to keep passing them to `merge` when the DocumentMapper has already been told about them.

Reply:
yes they ought to stay the same :) I see your point. these additional arguments do clutter the API, I just find it odd that we need to keep a reference to them only because it is needed for when merge is called, which is called from a place that has all this available too. I can be convinced though. I think not holding on to MapperService is the primary goal, hence I can go back and remove those arguments from merge.